### PR TITLE
Simplify testclusters, don't allow cross project clusters

### PR DESCRIPTION
**Pull request by alpar-t**
That static maps that hold references run a risk of causing memory leaks in the daemon. 
With this PR we introduce the limitation of not being able to use clusters defined in other projects and thus simplify this setup.

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: We can ditch the `synchronized` block here since this is being done at configuration time.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: This reference is never read or written from more than a single thread so there's no need for it to be marked `volatile`.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: Really? Do you have an example of the error handy?

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: This could be simplified a bit using `Map.compute()`. I'll leave that to your discretion.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: The comment below isn't applicable anymore.

> The downside is that with multi project builds if that other  task is in a different
> project and executing right now, we may terminate the cluster while it's running it.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: Similarly, could use `Map.compute()` here.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: I don't believe we need to use thread-safe collections here at all due to the single project limitation as configuration is serial and we don't execute multiple tasks in the same project in parallel.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: This seems pretty unnecessary. I'm certain there are already messages in the INFO log indicating the build has finished.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: I think this is going to cause memory leak issues in the daemon as we now have a global lister which holds a reference to this plugin and everything with it. We should either remove this an accept that some builds that terminate awkwardly might leave dangling processes or ensure we explicitly remove this listener.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: If we keep the JVM shutdown hook I don't think there's anything to prevent this from being called twice and in-turn, potentially calling `stop()` on our clusters multiple times.

**Review by mark-vieira — CHANGES_REQUESTED** [review]
A few comments. Mostly we need to address the issue with the shutdown hook because I think this is similarly going to cause daemon memory issues over time.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: I haven't saved it, but it's  saying exactly that, that passing a lambda here doesn't work. Probably has to do with some internal limitation. I added the comment because already bit me twice as IDEA suggests it could be a lambda.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: that should be fine because we still synchronize on `runningClusters` so there might be multiple calls to `shutDownAllClusters` but not to `stop` on the actual clusters

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: Excellent find! I made changes to remove the listener when the build completes.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: Interesting. SAM-coercion should kick in here.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: Synchonizing `runningClusters` won't help if we never remove them from the collection. We call `stop()` but we never actually clear out this collection so the subsequent call will call `stop()` again.

**Review by mark-vieira — APPROVED** [review]
Changes LGTM, assuming we can sort out the weird compile failure.

**Pull request merged**
